### ISSUE TITLE
Argo API returning "Error: Request entity too large: limit is 3145728" when workflow is huge

### ISSUE THREAD

**Issue by fangpenlin** [argoproj/argo#3738]
## Summary 

The argo API should be able to return payload bigger than 3MB given if the whole workflow graph is huge

## Diagnostics

v2.10.0-rc6

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: reproduce-bug-
spec:
  # keep workflows for 1d (86,400 seconds)
  ttlStrategy:
    secondsAfterCompletion: 86400
  # delete all pods as soon as they complete
  podGC:
    strategy: OnPodCompletion
  serviceAccountName: argo-workflow
  entrypoint: start
  arguments:
    parameters:
    - name: index
      value: "0000"

  templates:
  - name: start
    inputs:
      parameters:
      - name: index
    dag:
      tasks:
      - name: first-step
        template: first-step
        arguments:
          parameters:
          - name: index
            value: "{{inputs.parameters.index}}"
      - name: second-step
        dependencies: [first-step]
        template: second-step
        arguments:
          parameters:
          - name: index
            value: "{{inputs.parameters.index}}"
          - name: chunk
            value: "{{item}}"
        withParam: "{{tasks.first-step.outputs.result}}"
      - name: third-step
        dependencies: [second-step]
        template: third-step
        arguments:
          parameters:
          - name: index
            value: "{{inputs.parameters.index}}"
      - name: forth-step
        dependencies: [third-step]
        template: start
        arguments:
          parameters:
          - name: index
            value: "{{tasks.third-step.outputs.result}}"
        when: "{{tasks.third-step.outputs.result}} != EXIT"

  - name: first-step
    inputs:
      parameters:
      - name: index
    # allow to work for 5 minutes
    activeDeadlineSeconds: 300
    retryStrategy:
      limit: 5
      retryPolicy: "Always"
      backoff:
        duration: "15s"
        factor: 2
    script:
      image: python:3
      command: [bash]
      source: |
        set -e
        python3 -c 'import os, json; print(json.dumps([str(i) for i in range(300)]))'
      env:
      - name: LC_ALL
        value: C.UTF-8
      - name: LANG
        value: C.UTF-8
      resources:
        requests:
          cpu: 250m

  - name: second-step
    inputs:
      parameters:
      - name: index
      - name: chunk
    # allow to work for 30 minutes
    activeDeadlineSeconds: 1800
    retryStrategy:
      limit: 2
      retryPolicy: "Always"
      backoff:
        duration: "15s"
        factor: 2
    script:
      image: python:3
      command: [bash]
      source: |
        set -e
        echo hi {{$inputs.parameters.index}} {{$inputs.parameters.chunk}}
      env:
      - name: LC_ALL
        value: C.UTF-8
      - name: LANG
        value: C.UTF-8
      resources:
        requests:
          cpu: 100m

  - name: third-step
    inputs:
      parameters:
      - name: index
    # allow to work for 5 minutes
    activeDeadlineSeconds: 300
    script:
      image: python:3.7.8
      command: [python3]
      source: |
        current_index = int("{{inputs.parameters.index}}")
        next_index = current_index + 1
        if next_index>= 1024:
          print("EXIT")
        else:
          print(f"{next_index:04d}")
      env:
      - name: LC_ALL
        value: C.UTF-8
      - name: LANG
        value: C.UTF-8
      resources:
        requests:
          cpu: 100m


```

When you are trying to run a workflow that has very deep recursive times and very wide repeating tasks in the loop. Given my example, it's 1024 * 300, i.e around 300k tasks in this workflow. After a while, when enough tasks processed by argo, and you can open the argo web UI to see what's going on in the workflow page. You will notice error like this

```
Error: Request entity too large: limit is 3145728
```

Tried argo command like

```
argo get @latest
```

And the response could be something like this

```
Name:                reproduce-bug-2dr4r
Namespace:           default
ServiceAccount:      argo-workflow
Status:              Error
Message:             Request entity too large: limit is 3145728
Conditions:
 Completed           True
Created:             Tue Aug 11 02:41:06 -0700 (8 hours ago)
Started:             Tue Aug 11 02:41:06 -0700 (8 hours ago)
Finished:            Tue Aug 11 08:07:10 -0700 (3 hours ago)
Duration:            5 hours 26 minutes
```

I guess argo's http server for HTTP have a very small request size limit set at 3MB. I looked into the config but didn't found any option for controlling HTTP response size, did I miss something? Or maybe it makes sense to add an option for controlling this value to help serving big workflows?

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

**Comment by alexec** [comment]
You workflow is too large to fit into etcd. We have a specific feature to deal with this: node status offloading:

https://argoproj.github.io/argo/offloading-large-workflows/

**Comment by alexec** [comment]
Appears to work as expected. Removed `bug` label.

**Comment by eestewart** [comment]
> https://argoproj.github.io/argo/offloading-large-workflows/

Thanks.  I'll give this a try.

**Comment by fangpenlin** [comment]
@alexec ah, no wonder. the error message is a bit misleading though. I will turn on persistence / off load and try again.

the reproducing sample code I provided is simplified from my actual workflow, I didn't have time to run it and see if it can reproduce, I guess some part I removed in the code makes it not big enough to trigger the same issue.

anyway, thanks for the help, will update if the problem is gone.

**Comment by fangpenlin** [comment]
So, I ran with persistence option on with off loading large workflow on, doesn't seem to see the same problem. However, the web UI failed to load at some point, and my PostgreSQL seems like crashing. Also it's just too painful to have thousands of jobs in one workflow. So I turn the whole workflow into [Workflow of workflows](https://argoproj.github.io/argo/workflow-of-workflows/). So far it seems like working beautifully. Closing this issue for now, if people running into similar problem, I would recommend workflow of workflows instead of crazy size workflow.
